src/sys/socket/sockopt.rs

```diff
@@ -152,6 +153,10 @@ macro_rules! sockopt_impl {
         sockopt_impl!(Both, $name, $level, $flag, usize, GetUsize, SetUsize);
     };
 
+    (Both, $name:ident, $level:path, $flag:path, usize) => {
+        sockopt_impl!(Both, $name, $level, $flag, CString, GetCString, SetCString);
```

Comment:
I think `OsString` would be more appropriate than `CString`.

Reply:
I'm not sure of the difference between `OsString` and `CString`, and which is the best in this situation.

Reply:
`CString` should be used when you're interfacing with C libraries.  `OsString` should be used when you're interfacing with the operating system.  On Unix, they're usually the same thing.  But since `setsockopt` is provided by the OS, we should use `OsString` instead.

Reply:
Ok thank you.
However I cannot use `OsString` because I can't get ptr on it.

Reply:
Yes, but self.val is parametrized so I can't call `as_bytes()`.
```
mismatched types
expected &[u8], found type parameter
note: expected type `&[u8]` found type `T`
```

Reply:
Oh, I thought you were talking about the `as_ptr` call in `SetCString`.  Where does `OsString` cause a problem?

Reply:
Oh no sorry I mis explained. It's in `GetOsString`, I can't call `OsStr::from_bytes` because generic T has not conditions to be bytes. 
```
/// Getter for a `OsString` value.
struct GetOsString<T> {
    len: socklen_t,
    val: T,
}

unsafe impl<T> Get<OsString> for GetOsString<T> {
   ...
    unsafe fn unwrap(mut self) -> OsString {
        OsStr::from_bytes(self.val).to_owned()
    }
```

Reply:
You probably need to add a where guard to T: `T: AsRef<[u8]>`.

Reply:
Thank you. It's done but with OsString, `getsockopt(socket, sockopt::TcpCongestion)` returns an OsString with "reno\0\0\0\0\0\0\0\0\0\0\0\0" (with CString it's cropped to "reno").